### DISCUSSION TITLE
Adding Comments to Docusaurus Sites | AlexFornuto.com

### DISCUSSION THREAD

**Discussion by giscus** [alexfornuto/alexfornuto.com#6] [Announcements]
# Adding Comments to Docusaurus Sites | AlexFornuto.com

After recently moving my site from WordPress to Docusaurus, I decided I wanted a comments feature. Not because I got many comments before, nor have I written here in years (something I hope to write about soon), but for the challenge.

https://alexfornuto.com/blog/2024/04/01/giscus

**Comment by Eliav2** [comment]
This is wrong and does not work as expected. if you want comments on blog pages then `BlogPostItem` is not the right compnenet to swizzle. swizzeling BlogPostItem would also make a comments section around each blog preview in blogs page.

**Reply by Eliav2** [reply]
the right component to swizzle is `BlogPostPage` and you should add `<GiscusComponent />` right under BlogPostItem in `BlogPostPageContent`. like so:
```tsx
// src/theme/BlogPostPage/index.js

import React from 'react';
import clsx from 'clsx';
import { HtmlClassNameProvider, ThemeClassNames } from '@docusaurus/theme-common';
import { BlogPostProvider, useBlogPost } from '@docusaurus/theme-common/internal';
import BlogLayout from '@theme/BlogLayout';
import BlogPostItem from '@theme/BlogPostItem';
import BlogPostPaginator from '@theme/BlogPostPaginator';
import BlogPostPageMetadata from '@theme/BlogPostPage/Metadata';
import TOC from '@theme/TOC';
import Unlisted from '@theme/Unlisted';
import GiscusComponent from '@site/src/components/GiscusComponent';
function BlogPostPageContent({ sidebar, children }) {
    const { metadata, toc } = useBlogPost();
    const { nextItem, prevItem, frontMatter, unlisted } = metadata;
    const {
        hide_table_of_contents: hideTableOfContents,
        toc_min_heading_level: tocMinHeadingLevel,
        toc_max_heading_level: tocMaxHeadingLevel,
    } = frontMatter;
    return (
        <BlogLayout
            sidebar={sidebar}
            toc={
                !hideTableOfContents && toc.length > 0 ? (
                    <TOC toc={toc} minHeadingLevel={tocMinHeadingLevel} maxHeadingLevel={tocMaxHeadingLevel} />
                ) : undefined
            }
        >
            {unlisted && <Unlisted />}
            <BlogPostItem>{children}</BlogPostItem>
            <GiscusComponent />
            {(nextItem || prevItem) && <BlogPostPaginator nextItem={nextItem} prevItem={prevItem} />}
        </BlogLayout>
    );
}
export default function BlogPostPage(props) {
    const BlogPostContent = props.content;
    return (
        <BlogPostProvider content={props.content} isBlogPostPage>
            <HtmlClassNameProvider
                className={clsx(ThemeClassNames.wrapper.blogPages, ThemeClassNames.page.blogPostPage)}
            >
                <BlogPostPageMetadata />
                <BlogPostPageContent sidebar={props.sidebar}>
                    <BlogPostContent />
                </BlogPostPageContent>
            </HtmlClassNameProvider>
        </BlogPostProvider>
    );
}

```

**Reply by alexfornuto** [reply]
@Eliav2 Thanks for letting me know! The original content of this blog post _did_ [mention](https://github.com/alexfornuto/alexfornuto.com/commit/a912d0879044bce4b945ec603cdf3653a83123f8#diff-b9ce1e6ac01e84cb68a3181daeeaf00ccef43c74189ea4f03e234ff768d7a457R72) that the posts would end up on the preview page, so it was as expected for how it was described.

In any case, you identified the correct component to add the comments to, and I thank you. I've updated my site in #9 and the contents of this post in #10 to match.

**Comment by gaelhameon** [comment]
Thanks for this !
A simpler way to avoid showing the comments around each preview in the preview page is to use the `isBLogPostPage` value provided by the `useBlogPost()` hook. 

This allows using a "wrap-swizzling" (like you did in your first version) instead of an "eject-swizzling", and gives a much simpler swizzled component: 

```javascript
import React from 'react';
import BlogPostItem from '@theme-original/BlogPostItem';
import { useBlogPost } from '@docusaurus/theme-common/internal';
import GiscusComponent from '@site/src/components/GiscusComponent';

export default function BlogPostItemWrapper(props) {
  const { isBlogPostPage } = useBlogPost();
  return (
    <>
      <BlogPostItem {...props} />
      {isBlogPostPage && (
        <div >
          <hr />
          <GiscusComponent />
        </div>
      )}
    </>
  );
}
```

Another nice addition is to use `useDocusaurusContext` to retrieve the currentLocale in use this in the GiscusComponent: 

```
import React from 'react';
import Giscus from '@giscus/react';
import { useColorMode } from '@docusaurus/theme-common';
import useDocusaurusContext from '@docusaurus/useDocusaurusContext';

export default function GiscusComponent() {
  const { colorMode } = useColorMode();
  const { currentLocale } = useDocusaurusContext().i18n;
  return (
    <Giscus
      repo="repo"
      repoId="id"
      category="Announcements"
      categoryId="id"
      mapping="pathname"
      strict="0"
      reactionsEnabled="1"
      emitMetadata="0"
      inputPosition="top"
      theme={colorMode}
      lang={currentLocale}
      loading="lazy"
      crossorigin="anonymous"
      async
    />
  );
}
```

And finally, there is a typo in your post: 

> Define a new GiscusComponent component at src/components/GiscusComponents/index.js:

Based on the rest of your post, there is no "s" at /GiscusComponentS/index.js

**Comment by norahsakal** [comment]
Thanks for sharing this detailed walkthrough, really appreciate it!

Just a tiny typo I found (but hey that's what this comment section implementation is for, right? 😉):

You missed an **"s"** when importing GiscusComponent<ins>**s**</ins> from `src/components`:

It should be:

```
import GiscusComponent from '@site/src/components/GiscusComponents';
```

Thanks again!

**Comment by ghost** [comment]
thnk you

**Comment by mj-5** [comment]
Thank you.
But I have a problem, when I try to add some text comment and click `Comment`, it's freeze, and it doesn't record/saved. same problem when I try to add reactions.

**Comment by morewings** [comment]
Thanks for the tutorial.

Can you explain this?
```
      term="specific-term" //If you didn't select "Discussion title contains a specific term", omit.
```
